docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,270 @@
+labels of the `PackageRevision` available as values that may be controlled
+during the creation of the `PackageRevision`. This can assist in additional
+automation workflows.
+
+## Introducing Variance
+Just cloning is not that interesting, so the `PackageVariant` resource also
+allows you to control various ways of mutating the original package to create
+the variant.
+
+### Package Context[^notimplemented]
+Every kpt package that is designated `--for-deployment` will contain a
+ConfigMap called `kptfile.kpt.dev`. Kpt (or Porch) will automatically add a
+key `name` to the ConfigMap data, with the value of the package name. This
+ConfigMap can then be used as input to functions in the Kpt function pipeline.
+
+This process holds true for package revisions created via the package variant
+controller as well. Additionally, the author of the `PackageVariant` resource
+can specify additional key-value pairs to insert into the package
+context.
+
+### KRM Function Calls[^notimplemented]
+TODO(johnbelamaric): describe adding a KRM function pipeline to the
+PackageVariant to allow arbitrary mutations
+- question: should these allow adding to the package pipeline a la `--save`? Is
+  that a *separate* pipeline?
+
+### Configuration Injection[^pdc]
+Adding values to the package context, or executing functions with their
+configuration listed in the `PackageVariant` works for values that are under
+control of the author of the `PackageVariant` resource. However, in more
+advanced use cases, we may need to specialize the package based upon other
+contextual information. This particularly comes into play when the user
+deploying the workload does not have direct control over the context in which
+it is being deployed. For example, one part of the organization may manage the
+infrastructure - that is, the cluster in which we are deploying the workload -
+and another part the actual workload. We would like to be able to pull in inputs
+specified by the infrastructure team automatically, based the cluster to which
+we are deploying the workload, or perhaps the region in which that cluster is
+deployed.
+
+To facilitate this, the package variant controller can "inject" configuration
+directly into the package. This means it will use information specific to this
+instance of the package to lookup a resource in the Porch cluster, and copy that
+information into the package. Of course, the package has to be ready to receive
+this information. So, there is a protocol for facilitating this dance:
+- Packages may contain resources annotated with `kpt.dev/config-injection`
+- Usually, these will also be `config.kubernetes.io/local-config` resources, as
+  they are likely just used by local functions as input. But this is not
+  mandatory.
+- The package variant controller will look for any resource in the Kubernetes
+  cluster matching the Group, Version, and Kind of the package resource, and
+  satisfying the *injection selector*.
```

Comment:
To @mortent's point above, perhaps we should inject a `local-config` version of the targeting resource, either automatically or with a simple injection selector - that is, without requiring a pre-existing local object in the package.

Reply:
Actually this is not something I want to do. I'd rather require the package advertise the ways it accepts injections (ie, the GVKs it understands), and only insert those. This keeps the separation of concerns cleaner; the package doesn't build in awareness of the context in which it's being deployed. For example, I don't want a package to accept a Porch Repository resource just because that happens to be the targeting mechanism. If it needs a value from, say, an annotation on the repository, then we should provide a mechanism to inject that value. If, on the other hand, it needs a whole resource of some sort, then we need to define the right injection selectors in the PV/PVS so that the controller can find it and inject it.

Reply:
Current thinking is this (will update the proposal soon):

1) There must be an in-package object with config-injection annotation
2) GVK of in-package and in-cluster object must match
3) PV will contain an ordered list of selectors, first match wins
4) At least for now selectors will be an explicit name (ie, PV will look for an object with that name, in the same namespace as the PV, with the same GVK as the in-package object)
5) These object must have a `spec` field, and only the `spec` will be copied from the in-cluster object to the in-package object
6) the in-package object will be annotated with the name of the object that was selected (recall we can have multiple selectors)
7) the in-package object will be annotated with a hash of the contents of the spec field (so that the PV controller can evaluate whether the spec has been changed in either object during later reconciliations, and fix config drift)
8) PVS will add a "nameFrom" or something similar, that will resolve to the explicity "name" fields in the selectors during fan out. This allows the PVS to, for example, set the injector name based upon the target name or another target field, or even an annotation or label defined on the target.
9) maybe someday we will expand these selectors (both injection on PV and derivation of that in PVS) to label selectors, etc., but right now there is no clear use case for that
10) Regardless of whether any injection selectors are defined, the PV controller will create package conditions corresponding to each resource tagged with `config-injection`. If the PV controller successfully injects the config, that condition gets set to `True`, otherwise `False` with an appropriate message.

Reply:
The selectors should be able to optionally include GVK. This is way to constrain the injections that will be done. That is, if the package has 10 different objects with `config-injection` annotation, the PV could say it only wants to replace certain GVKs. This also allows a little better control. So, for example, if you had in your cluster

GVK1 foo
GVK1 bar
GVK2 foo
GVK2 bar

you could define a list of selectors like:

 - GVK1 foo
 - GVK2 var

without being able to include GVK in the selector, you would always get "foo" for every injection point. with these selectors we get foo for GVK1 and bar for GVK2.

Reply:
We probably want to differentiate the condition of whether injection succeeded or not based on if the GVK of the in-package resource matches a selector. This allows the PV author to say they only want to "fill in" particular injection points, and leave the others alone (for example, to override a default or not).

This begs the question though - should we have a selector for the in-package resources of interest, and then a separate matcher-selector to match each of those to in-cluster objects. That's sounding complicated.

Reply:
Another option: the `config-injection` annotation can be used to flag whether or not failed injection should create a False condition. In package deployment controller, it is just true/false. But we can do better - for example, if the annotation is "required", then a failed injection results in a False condition. If it is "optional", then even if we fail to find a match, we create the condition as True (and just put in the message that it was left as the default).

Reply:
Updated to reflect this.